drivers/CpuUid.h

```diff
@@ -0,0 +1,128 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CPUUID_H
+#define MBED_CPUUID_H
+
+#include "platform/platform.h"
+
+#if defined(DEVICE_CPUUID) || defined(DOXYGEN_ONLY)
+
+#ifdef MBED_CPU_UID_STR_SIZE_MAX
+#define CPU_UID_STRING_BUFFER_SIZE  MBED_CPU_UID_STR_SIZE_MAX
+#else
+#define CPU_UID_STRING_BUFFER_SIZE (MBED_CPU_UID_SIZE * 2 + 1)
+#endif
+
+namespace mbed {
+/** \addtogroup drivers */
+
+/** CPU UID reader class
+ *
+ * @note Synchronization level: Interrupt safe
+ * @ingroup drivers
+ */
+class CpuUid {
+public:
+    /** CpuUid constructor
+     */
+    CpuUid();
+    
+    /** Get size of CPU UID in bytes
+     * 
+     * @return Size of device's CPU UID in bytes
+     */
+    int size()
+    {
+        return MBED_CPU_UID_SIZE;
+    }
+    
+    /** Get CPU UID data pointer
+     * 
+     * @return Pointer to uid data buffer
+     */
+    const uint8_t *data()
```

Comment:
Why are we using uint8_t* to identify the device when most devices use one or more 32-bit words for identification? What's the use case here?

Reply:
Use case would be in our case: unique device identification in production ... and for support issues afterwards.

Reply:
I see your point. But the problem with string representation of Serial Number/Device-ID is it can become problematic if we want to adopt Unicode support in future or if we are sending this to other entities which might be using different string format ( E.g: UTF8, UTF16 etc ) and may cause inter-operability issues/may need conversions again. What I suggest is, instead of storing the UID as string, we can store the SerialNumber/DeviceID in its raw form(with some metadata if required like length, structure etc) and then provide member functions(E.g: CpuUid.GetIdAsString()) to retrieve it in string format. And the buffer for the output will be provided by the caller so CpuUid class doesn't have to manage the string buffers.

Reply:
When Unicode support gets added to mbed, it's still time to think about that ;)

I thought about doing another refactoring of the driver, so the uint8_t* representation of the uid won't get "cached" as static member, and instead being retrieved from hal each time called, and written to a local buffer provided by the caller.
That way no buffer at all would be necessary in the driver, and since the reading of the uid is usually some copy&shift work directly from registers, this is not really a performance issue.

This could be done also for the vendor provided string representation, but as you point out in the comment below, string operations can be lengthy - that's why I liked @bremoran 's idea with the static buffer...

Reply:
We don't need to *store* the SerialNumber/DeviceID anywhere. It's just fine to fetch it on-demand. If you really must store it, please ensure it's a static class member.